Assets/Scripts/Puzzle.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+/// <summary>
+/// A puzzle is an object connected to a switch or group of switches.
+/// It has a solution, a string formed from a combination of each switch's character.
+/// When the player activates the switches in a sequence which matches this solution,
+/// the puzzle's state changes to allow the player to progress.
+/// </summary>
+public abstract class Puzzle : MonoBehaviour
+{
+    [SerializeField]
+    [Tooltip("The solution to the puzzle.\n This should be a combination " +
+        "of the connected switches' solutionEntries.")]
+    private string solution = "1";
+
+    /// <summary>
+    /// The current attempted solution
+    /// </summary>
+    private string currentSolutionAttempt_useProperty;
+
+    /// <summary>
+    /// Property to reference the current solution attempt. If the solution
+    /// is completely correct, the puzzle performs its "solved" function.
+    /// Otherwise, it resets the attempt and invokes SolutionIsIncorrect
+    /// </summary>
+    public virtual string CurrentSolutionAttempt
+    {
+        get { return currentSolutionAttempt_useProperty; }
+        set
+        {
+            currentSolutionAttempt_useProperty = value;
+            if (CheckIfSolutionMatchesSoFar())
+            {
+                if (CurrentSolutionAttempt.Length == solution.Length)
+                    DoPuzzleSolved();
+            }
+            else
+            {
+                currentSolutionAttempt_useProperty = "";
+                if (SolutionIsIncorrect != null)
+                    SolutionIsIncorrect.Invoke(this);
+            }
+        }
+    }
+
+    /// <summary>
+    /// An event to invoke when the solution attempt fails
+    /// </summary>
+    public static event System.Action<Puzzle> SolutionIsIncorrect;
```

Comment:
Events should have past tense names to indicate what just happened (e.g. ButtonClicked). So, something like IncorrectSolutionAttempted or  PuzzleFailed whatever makes sense to you. Then you make event handlers like "OnPuzzleFailed()"